flang/lib/Lower/ConvertType.cpp

```diff
@@ -411,185 +211,86 @@ class TypeBuilder {
+    using TC =
+        Fortran::evaluate::Type<Fortran::common::TypeCategory::Character, Kind>;
+    auto designator = Fortran::evaluate::Fold(
+        foldingContext,
+        Fortran::evaluate::Expr<TC>{Fortran::evaluate::Designator<TC>{symbol}});
+    if (auto len = toInt64(std::move(designator.LEN())))
+      return *len;
+    return fir::SequenceType::getUnknownExtent();
   }
-
-  //===--------------------------------------------------------------------===//
-
-  mlir::MLIRContext *context;
-  const Fortran::common::IntrinsicTypeDefaultKinds &defaults;
-};
-
-/// NewTypeBuilder simplifies the expression type conversion by operating
-/// on evaluate::DynamicType and evaluate::Shape instead of going down the
-/// expression tree. It fixes the fact that genType("array") would
-/// return the scalar type and crash code using the result to dispatch between
-/// scalar and array handling.
-/// TODO: replace the previous expression type conversion with it or fix the
-/// current one (The old one is still used for DataRef and Designator).
-struct NewTypeBuilder {
-
-  mlir::Type gen(const Fortran::lower::SomeExpr &expr) {
-    auto dynamicType = expr.GetType();
-    if (!dynamicType)
-      llvm::report_fatal_error("lowering typeless expression type");
-    auto category = dynamicType->category();
-    if (category == Fortran::common::TypeCategory::Derived)
-      TODO("derived types lowering");
-    auto shapeExpr = Fortran::evaluate::GetShape(foldingContext, expr);
-    if (!shapeExpr)
-      TODO("implied shape expression type lowering");
-
-    auto baseType = TypeBuilder{context, foldingContext.defaults()}.genFIRTy(
-        category, dynamicType->kind());
-    if (category == Fortran::common::TypeCategory::Character) {
-      auto len = fir::SequenceType::getUnknownExtent();
-      if (auto constantLen = toInt64(dynamicType->GetCharLength()))
-        len = *constantLen;
-      fir::SequenceType::Shape shape{len};
-      translateShape(shape, std::move(*shapeExpr));
-      return fir::SequenceType::get(shape, baseType);
+  fir::SequenceType::Extent
+  getCharacterLength(const Fortran::semantics::Symbol &symbol) {
+    auto *type = symbol.GetType();
+    if (!type ||
+        type->category() != Fortran::semantics::DeclTypeSpec::Character ||
+        !type->AsIntrinsic())
+      llvm::report_fatal_error("not a character symbol");
+    int kind =
+        toInt64(Fortran::common::Clone(type->AsIntrinsic()->kind())).value();
+    switch (kind) {
+    case 1:
+      return getCharacterLengthHelper<1>(symbol);
+    case 2:
+      return getCharacterLengthHelper<2>(symbol);
+    case 4:
+      return getCharacterLengthHelper<4>(symbol);
     }
-    // LOGICAL, INTEGER, REAL, COMPLEX
-    fir::SequenceType::Shape shape{};
-    translateShape(shape, std::move(*shapeExpr));
-    if (!shape.empty())
-      return fir::SequenceType::get(shape, baseType);
-    return baseType;
+    llvm::report_fatal_error("unknown character kind");
```

Comment:
I think we could even use the weaker `llvm_unreachable` here, as the KIND values are hard-coded into the front-end.